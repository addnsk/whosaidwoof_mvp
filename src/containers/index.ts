export * from "./Header";
export * from "./Theme";
export * from "./SelectList";
