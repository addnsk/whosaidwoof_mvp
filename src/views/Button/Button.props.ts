import * as React from "react";

type ButtonAppearance = "primary" | "outline" | "bubble" | "bubbleDark";

export interface IButtonProps
  extends React.DetailedHTMLProps<
    React.ButtonHTMLAttributes<HTMLButtonElement>,
    HTMLButtonElement
  > {
  appearance: ButtonAppearance;
  caption: string;
}
