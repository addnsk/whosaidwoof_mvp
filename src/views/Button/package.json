{
  "name": "button-view",
  "version": "0.1.0",
  "private": true,
  "main": "Button.tsx"
}
