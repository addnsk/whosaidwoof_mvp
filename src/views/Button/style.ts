import styled, { css } from "styled-components";
import { IButtonProps } from "./Button.props";

const appearanceMapper = (appearance: string) => {
  switch (appearance) {
    case "primary":
      return css`
        width: 185px;
        height: 40px;
        border: none;
        border-radius: 15px;
        background-color: ${({ theme }) => theme.colors.light};
        color: ${({ theme }) => theme.colors.dark};
        &:active {
          background-color: ${({ theme }) => theme.colors.accentLight};
        }
      `;
    case "outline":
      return css`
        width: 185px;
        height: 40px;
        border-radius: 15px;
        border: 1px solid ${({ theme }) => theme.colors.light};
        background-color: transparent;
        &:active {
          border: 1px solid ${({ theme }) => theme.colors.accentLight};

          & * {
            color: ${({ theme }) => theme.colors.accentLight};
          }
        }
      `;
    case "bubble":
      return css`
        width: 44px;
        height: 44px;
        border-radius: 17px;
        border: 1px solid ${({ theme }) => theme.colors.light};
        background-color: transparent;
        &:active {
          border: 1px solid ${({ theme }) => theme.colors.accentLight};

          & * {
            color: ${({ theme }) => theme.colors.accentLight};
          }
        }
      `;
    case "bubbleDark":
      return css`
        width: 44px;
        height: 44px;
        border-radius: 17px;
        border: 1px solid ${({ theme }) => theme.colors.accentDark};
        background-color: transparent;
        color: ${({ theme }) => theme.colors.accentDark};
        &:active {
          border: 1px solid ${({ theme }) => theme.colors.light};

          & * {
            color: ${({ theme }) => theme.colors.light};
          }
        }
      `;
    default:
      return css`
        width: fit-content;
        height: fit-content;
        border: none;
        background-color: ${({ theme }) => theme.colors.transparent};
      `;
  }
};

export const Division = styled.button<IButtonProps>`
  display: flex;
  justify-content: center;
  align-items: center;
  margin: 0;
  padding: 0 8px;
  box-sizing: border-box;
  cursor: pointer;

  outline: none;
  user-select: none;

  -webkit-tap-highlight-color: rgba(0, 0, 0, 0);

  ${({ theme }) => theme.typography.buttonText}
  ${({ appearance }) => appearanceMapper(appearance)}

  &:disabled {
    pointer-events: none;
    opacity: 0.5;
  }

  &:focus {
    box-shadow: 0 0 0 1px ${({ theme }) => theme.colors.accentLight};
  }
`;
