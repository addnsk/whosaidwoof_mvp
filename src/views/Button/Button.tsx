import * as React from "react";
import { IButtonProps } from "./Button.props";
import { Division } from "./style";

export const Button: React.FC<IButtonProps> = React.memo(
  ({ appearance, caption, onClick, disabled, type }): JSX.Element => {
    return (
      <Division
        type={type}
        appearance={appearance}
        caption={caption}
        onClick={onClick}
        disabled={disabled}
      >
        {caption && <span>{caption}</span>}
      </Division>
    );
  }
);
