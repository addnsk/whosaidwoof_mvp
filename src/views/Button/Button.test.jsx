import "@testing-library/jest-dom";
import { render, screen } from "@testing-library/react";
import { Theme } from "containers";
import { Button } from "./Button";
import userEvent from "@testing-library/user-event";

const onClick = jest.fn();

describe("Button component", () => {
  it("Button renders with full props", () => {
    render(
      <Theme>
        <Button
          type="button"
          appearance="primary"
          caption="fun button"
          onClick={onClick}
        />
      </Theme>
    );
    expect(screen.getByRole("button")).toBeInTheDocument();
    expect(screen.getByRole("button")).toHaveStyle({
      backgroundColor: "#FFFFFF",
      color: "#2A2F44",
      width: "185px",
    });

    expect(screen.getByText(/fun button/i)).toBeInTheDocument();

    expect(screen.getByRole("button")).toMatchSnapshot();
  });

  it("Button renders with some default props and children", () => {
    render(
      <Theme>
        <Button caption="fun button">Goodbye</Button>
      </Theme>
    );
    expect(screen.getByRole("button")).toBeInTheDocument();
    expect(screen.getByRole("button")).toHaveStyle({
      backgroundColor: "transparent",
      width: "fit-content",
    });
    expect(screen.getByText(/fun button/i)).toBeInTheDocument();
    expect(screen.queryByText(/goodbye/i)).toBeNull();

    expect(screen.getByRole("button")).toMatchSnapshot();
  });

  it("Button renders without props", () => {
    render(
      <Theme>
        <Button />
      </Theme>
    );
    expect(screen.getByRole("button")).toBeInTheDocument();
    expect(screen.getByRole("button").firstChild).toBeNull();

    expect(screen.getByRole("button")).toMatchSnapshot();
  });

  it("Button responds to onClick", () => {
    render(
      <Theme>
        <Button
          type="button"
          appearance="primary"
          caption="fun button"
          onClick={onClick}
        />
      </Theme>
    );

    userEvent.click(screen.getByRole("button"));

    expect(onClick).toHaveBeenCalledTimes(1);
  });

  it("Button doesn`t respond to onClick if it`s disabled", () => {
    render(
      <Theme>
        <Button
          type="button"
          disabled
          appearance="primary"
          caption="fun button"
          onClick={onClick}
        />
      </Theme>
    );

    userEvent.click(screen.getByRole("button"));

    expect(onClick).toHaveBeenCalledTimes(0);
    expect(screen.getByRole("button")).toHaveStyle({
      opacity: 0.5,
    });
  });
});
