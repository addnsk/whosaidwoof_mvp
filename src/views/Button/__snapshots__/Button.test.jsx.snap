// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component Button renders with full props 1`] = `
<button
  class="sc-bdfBQB eWbZmV"
  type="button"
>
  <span>
    fun button
  </span>
</button>
`;

exports[`Button component Button renders with some default props and children 1`] = `
<button
  class="sc-bdfBQB bzyqGh"
>
  <span>
    fun button
  </span>
</button>
`;

exports[`Button component Button renders without props 1`] = `
<button
  class="sc-bdfBQB bzyqGh"
/>
`;
